### PR TITLE
feat(tooling): add-logic-to-generate-changelogs

### PR DESCRIPTION
**Pull request by Shreyas281299**
# COMPLETES #[SPARK-532257](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-532257)

## This pull request addresses
Added a new command in the package-tools to create changelogs.
The command is run on deploy after npm publish has already happened.

## by making the following changes
When changelog mehtod is called a changelog file is created or the existing changelog file is updated in the documentation branch.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [x] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

Updated pull-request.yml to run the changelog command in my fork. 

- https://github.com/Shreyas281299/webex-js-sdk/actions/runs/9807909431/job/27082945372 - Here is the github action that created the following commit
-https://github.com/Shreyas281299/webex-js-sdk/tree/documentation - Here is the commit that added the changelog. 

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/tools/package/src/commands/changelog/changelog.constants.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  name: 'changelog',
+  description: 'Update the changelog',
+  options: [
+    {
+      description: 'Packages to be added in changelog',
+      name: 'packages',
+      type: 'string...',
+    },
+    {
+      description: 'Tag to update this local package with on npm',
```
Comment (mkesavan13) [inline packages/tools/package/src/commands/changelog/changelog.constants.ts:19]: Could we please reword this?

File: packages/tools/package/src/commands/changelog/changelog.types.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+   * Details about the packages that are changed as part of the version.
+   */
+  [packageName: string]: string;
+}
+
+export interface ChangelogEntry {
+  [packageName: string]: {
+    [version: string]: {
+      commits: CommitDetails;
+      alongWith: AlongWithData;
```
Comment (mkesavan13) [inline packages/tools/package/src/commands/changelog/changelog.types.ts:43]: We also need the `published_date` type added. It should be a Unix timestamp

File: packages/tools/package/test/module/changelog/changelog.fixtures.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+const packagesData = {
+  webex: {
+    data: {
+      location: '/Users/shreysh2/Desktop/Repo/webex-js-sdk/packages/webex',
```
Comment (mkesavan13) [inline packages/tools/package/test/module/changelog/changelog.fixtures.js:4]: Why is your system details here? Should we consider replacing it with a generic one?

File: packages/tools/package/test/module/changelog/changelog.fixtures.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        },
+      },
+      version: {
+        major: 3, minor: 3, patch: 1, release: 6, tag: 'next',
+      },
+    },
+  },
+  '@webex/package-tools': {
+    data: {
+      location: '/Users/shreysh2/Desktop/Repo/webex-js-sdk/packages/tools/package',
```
Comment (mkesavan13) [inline packages/tools/package/test/module/changelog/changelog.fixtures.js:22]: Same here

File: packages/tools/package/src/commands/changelog/changelog.constants.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * @public
+ */
+const CONFIG: CommandsConfig = {
+  name: 'changelog',
+  description: 'Update the changelog',
+  options: [
+    {
+      description: 'Packages to be added in changelog',
+      name: 'packages',
+      type: 'string...',
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.constants.ts:16]: Is it a typo? `string...`

File: packages/tools/package/src/commands/changelog/changelog.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+   */
+  handler: async (options: Options) => {
+    if (!options.packages) {
+      return Promise.resolve({});
+    }
+    const rootDir = process.cwd();
+
+    const tag = options.tag?.split('/').pop();
+    return Yarn.list()
+      .then((packageDetails) => packageDetails.map(
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.ts:37]: Since we have already declared the function as async, we can use `async-await` syntax

File: packages/tools/package/src/commands/changelog/changelog.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import { Package } from '../../models';
+import { Yarn } from '../../utils';
+import type { PackageConfig } from '../../models';
+
+import CONSTANTS from './changelog.constants';
+import type { Options } from './changelog.types';
+import { createOrUpdateChangelog } from './changelog.utils';
+
+/**
+ * @public
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.ts:14]: We can add description here

File: packages/tools/package/src/commands/changelog/changelog.types.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  commit: string;
+}
+interface CommitDetails {
+  /**
+   * Details about the commits added as part of the version.
+   */
+  [key: string]: string;
+}
+
+export interface AlongWithData {
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.types.ts:32]: Q. What does this `AlongWithData` means?

File: packages/tools/package/src/commands/changelog/changelog.utils.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+const fs = require('fs');
+const path = require('path');
+const { exec } = require('child_process');
+
+/**
+ * Function to execute shell command and capture output
+ * @param scriptPath - Path to the script which we want to run.
+ * @returns - Promise that resolves once the script is run and returns the output.
+ */
+export function runShellScript(scriptPath: string) {
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.utils.ts:13]: This seems generic function. We can check if this is defined somewhere else as well

File: packages/tools/package/src/commands/changelog/changelog.utils.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        resolve(stdout);
+      }
+    });
+  });
+}
+
+/**
+ * Function to get and formate alongWith object
+ * alongWith object has the information about packages that are modified `along with`
+ * a package
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.utils.ts:34]: Description could be a bit more structured

File: .github/workflows/deploy.yml [hunk 11]
```diff
@@ -254,7 +254,12 @@ jobs:
           git config user.email "${GIT_AUTHOR_EMAIL}"
           git config user.name "${GIT_AUTHOR_NAME}"
 
-      - name: Publish Docs Folder
+      - name: Updated changelog
```
Comment (sreenara) [inline .github/workflows/deploy.yml:257]:
```suggestion
      - name: Update changelog
```

File: .github/workflows/pull-request.yml [hunk 12]
```diff
@@ -168,7 +168,6 @@ jobs:
           path: '**/dist'
           key: dist-${{ env.rid }}
 
-
```
Comment (sreenara) [inline .github/workflows/pull-request.yml:171]: This change isn't required right? Can we revert it

File: packages/tools/package/src/commands/changelog/changelog.utils.ts [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+      } else {
+        // Process the stdout (output) as needed
+        resolve(stdout);
+      }
+    });
+  });
+}
+
+/**
+ * Function to get and formate alongWith object
```
Comment (sreenara) [inline packages/tools/package/src/commands/changelog/changelog.utils.ts:32]: Typo in the word format

File: packages/tools/package/package.json [hunk 14]
```diff
@@ -30,6 +30,7 @@
     "start": "node ./index.js",
     "test": "yarn test:style && yarn test:syntax && yarn test:integration && yarn test:coverage",
     "test:coverage": "yarn test:integration --coverage --reporters=\"jest-silent-reporter\"",
+    "test:unit": "jest",
```
Comment (sreenara) [inline packages/tools/package/package.json:33]: Why was this needed? UTs were running for the other commands in this package before right?

**Review by sreenara — CHANGES_REQUESTED** [review]
Please see the comments. Also, can we add these tests?
1. Where the changelog is already present and a new changelog is appended?
2. We move to a new version such as 3.4.0 or 4.0.0 and a new file is created?

File: packages/tools/package/package.json (hunk 14 above)
Comment (Shreyas281299) [reply packages/tools/package/package.json:33]: We run the `test:unit` command to run unit tests for all the packages but here we didn't have the command. Although nothing is broken because test:integration is running the unit test as you can see in the command below.

File: packages/tools/package/src/commands/changelog/changelog.types.ts (hunk 8 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.types.ts:32]: Details about the alongwithdata is given in line 33-36

File: packages/tools/package/src/commands/changelog/changelog.ts (hunk 7 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.ts:14]: I've followed the style with which comments are written in the tooling. The description is written inside the constant so here line 17-20 has the description

File: packages/tools/package/src/commands/changelog/changelog.ts (hunk 6 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.ts:37]: Removed async-await. Trying to keep all the tooling consistent with the .then approach

File: packages/tools/package/test/module/changelog/changelog.fixtures.js (hunk 4 above)
Comment (Shreyas281299) [reply packages/tools/package/test/module/changelog/changelog.fixtures.js:22]: Removed

File: packages/tools/package/test/module/changelog/changelog.fixtures.js (hunk 3 above)
Comment (Shreyas281299) [reply packages/tools/package/test/module/changelog/changelog.fixtures.js:4]: Removed.

File: packages/tools/package/src/commands/changelog/changelog.types.ts (hunk 2 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.types.ts:43]: Added

File: packages/tools/package/src/commands/changelog/changelog.constants.ts (hunk 1 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.constants.ts:19]: Updated.

File: packages/tools/package/src/commands/changelog/changelog.utils.ts (hunk 10 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.utils.ts:34]: Ive tried to make the description a bit more clear.

File: .github/workflows/deploy.yml (hunk 11 above)
Comment (Shreyas281299) [reply .github/workflows/deploy.yml:257]: Changed

File: .github/workflows/pull-request.yml (hunk 12 above)
Comment (Shreyas281299) [reply .github/workflows/pull-request.yml:171]: Removed

File: packages/tools/package/src/commands/changelog/changelog.utils.ts (hunk 13 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.utils.ts:32]: Fixed

File: packages/tools/package/src/commands/changelog/changelog.utils.ts (hunk 9 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.utils.ts]: Yes, there is an executor class that is used to run command. Using that class instead on the shell script.

**Comment by Shreyas281299** [comment]
> Please see the comments. Also, can we add these tests?
> 
>     1. Where the changelog is already present and a new changelog is appended?
> 
>     2. We move to a new version such as 3.4.0 or 4.0.0 and a new file is created?

Added test. These tests were missing I have added a new file with these test - changelog.utils.test

File: packages/tools/package/package.json (hunk 14 above)
Comment (sreenara) [reply packages/tools/package/package.json:33]: @Shreyas281299 let's remove the test:integration command then. It's not needed.

File: packages/tools/package/test/module/changelog/changelog.utils.test.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
+const changelogUtils = require('@webex/package-tools/dist/module/commands/changelog/changelog.utils');
+const changelogExecutor = require('@webex/package-tools/dist/module/commands/changelog/changelog.executor');
```
Comment (sreenara) [inline packages/tools/package/test/module/changelog/changelog.utils.test.js:2]: Why do we have to require this way with a complete path? Does relative path not work? Is there some error?

File: packages/tools/package/test/module/changelog/changelog.utils.test.js [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+      await changelogUtils.createOrUpdateChangelog(fixtures.packagesData, 'mock_commit_id');
+      expect(fs.readFileSync).toHaveBeenCalledWith(`./docs/changelog/v${filePath331}.json`);
+    });
+
+    it('dont read changelog data and create changelog file if version file doesnt exists', async () => {
+      fs.existsSync.mockReturnValue(false);
+      await changelogUtils.createOrUpdateChangelog(fixtures.packagesData, 'mock_commit_id');
+      expect(fs.readFileSync).not.toHaveBeenCalledWith();
+      expect(fs.mkdirSync).toHaveBeenCalledWith('./docs/changelog', {
+        recursive: true,
```
Comment (sreenara) [inline packages/tools/package/test/module/changelog/changelog.utils.test.js:42]: What's the filename of the new file? Are we not going to check that?

File: packages/tools/package/test/module/changelog/changelog.utils.test.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+      });
+    });
+
+    it('creates correct changelog data and writes to the respective changelog file', async () => {
```
Comment (sreenara) [inline packages/tools/package/test/module/changelog/changelog.utils.test.js:46]: I think we could reword this test. If I understand correctly, this test creates a new changelog file and fills it with the right content, right?

File: packages/tools/package/test/module/changelog/changelog.utils.test.js [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+      jest.spyOn(Date, 'now').mockReturnValue('123456789');
+      fs.readFileSync.mockReturnValue(fixtures.changelogData);
+      await changelogUtils.createOrUpdateChangelog(fixtures.packagesData, 'mock_commit_id');
+      const filePath000 = fixtures.packagesData[1].data.packageInfo.version
+        .split('-')[0]
+        .replace(/\./g, '_');
+      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
+      expect(fs.writeFileSync.mock.calls).toEqual([
```
Comment (sreenara) [inline packages/tools/package/test/module/changelog/changelog.utils.test.js:54]: Can we have a comment that explains these 2 calls we are expecting?

**Review by sreenara — CHANGES_REQUESTED** [review]
Please see the comments

File: packages/tools/package/test/module/changelog/changelog.utils.test.js (hunk 15 above)
Comment (Shreyas281299) [reply packages/tools/package/test/module/changelog/changelog.utils.test.js:2]: Yes. Im unable to import them relatively. I have used this also to maintain coding convention in packag-tools. This is exactly how utils have been tested in this package.

File: packages/tools/package/test/module/changelog/changelog.utils.test.js (hunk 16 above)
Comment (Shreyas281299) [reply packages/tools/package/test/module/changelog/changelog.utils.test.js:42]: Actually mkdirSync was used as a fallback to create the changelog folder if its not present. And it is fs.writeFileSync that creates or edits the files. So I have removed this method and the test.

File: packages/tools/package/test/module/changelog/changelog.utils.test.js (hunk 17 above)
Comment (Shreyas281299) [reply packages/tools/package/test/module/changelog/changelog.utils.test.js:46]: Yes. It creates the data, and either populates the existing file or creates a new file and puts the data their.

Comment (Shreyas281299) [reply packages/tools/package/test/module/changelog/changelog.utils.test.js]: I have updated the description. Please see

File: packages/tools/package/test/module/changelog/changelog.utils.test.js [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+      jest.spyOn(Date, 'now').mockReturnValue('123456789');
+      fs.readFileSync.mockReturnValue(fixtures.changelogData);
+      await changelogUtils.createOrUpdateChangelog(fixtures.packagesData, 'mock_commit_id');
+      const filePath000 = fixtures.packagesData[1].data.packageInfo.version
+        .split('-')[0]
+        .replace(/\./g, '_');
+      expect(fs.writeFileSync).toHaveBeenCalledTimes(2);
+      // writeFileSync will be called for each package
+      // Here, according to the fixtures first it will be called for webex
+      // and second, it will be called for @webex/package-tools
```
Comment (sreenara) [inline packages/tools/package/test/module/changelog/changelog.utils.test.js:47]:
```suggestion
      // writeFileSync will be called for each package
      // Here, according to the fixtures first it will be called for webex which is at version 3.3.1
      // and second, it will be called for @webex/package-tools at version 0.0.0
```

**Review by sreenara — APPROVED** [review]
Approved with a nitpick in the comment

**Review by sreenara — CHANGES_REQUESTED** [review]
Can we check why the UT is failing (1 failure)?
Coverage is also failing.

File: packages/tools/package/test/module/changelog/changelog.utils.test.js (hunk 19 above)
Comment (Shreyas281299) [reply packages/tools/package/test/module/changelog/changelog.utils.test.js:47]: Updated

**Comment by Shreyas281299** [comment]
> Can we check why the UT is failing (1 failure)? Coverage is also failing.

Coverage was failling because in test:coverage we were calling test:integration and since that command has been removed it was failling I have replaced test:integration with test:unit

UTs were failling because i didnt handle the existing changelog data in uts properly. Fixed that as well.

File: .github/workflows/deploy.yml [hunk 20]
```diff
@@ -254,7 +254,12 @@ jobs:
           git config user.email "${GIT_AUTHOR_EMAIL}"
           git config user.name "${GIT_AUTHOR_NAME}"
 
-      - name: Publish Docs Folder
+      - name: Update changelog
+        run: |
+          git checkout documentation docs/changelog/
+          yarn package-tools changelog --packages ${{ needs.generate-package-matrix.outputs.node-recursive }} --tag ${GITHUB_REF##*/} --commit ${{ github.event.before }}
+
+      - name: Update Changelog and Publish Docs Folder
```
Comment (rarajes2) [inline .github/workflows/deploy.yml:262]: Are we not following any convention for the name description here? Somewhere we are using PascalCase and somewhere it's normal sentence.

File: packages/tools/package/src/commands/changelog/changelog.ts (hunk 6 above)
Comment (rarajes2) [reply packages/tools/package/src/commands/changelog/changelog.ts:38]: I don't think we need to maintain consistency with thenable syntax. We should use async-await everywhere now, it's more readable.

File: packages/tools/package/src/commands/changelog/changelog.utils.ts [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Function to create or update changelog files
+ * @param packages - Details of the modified packages
+ * @param prevCommitId - commitId of the previous commit
+ */
+export async function createOrUpdateChangelog(packages: Package[], prevCommitId: string) {
+  Object.keys(packages).forEach(async (index: any) => {
+    const pkgName = packages[index].name;
+    // const {version} = packages[index];
+    // console.log(packages[index], version);
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.utils.ts:35]: We can remove all the commented lines

File: packages/tools/package/src/commands/changelog/changelog.utils.ts [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+    const { version } = packages[index];
+    const fileName = version.split('-')[0].replace(/\./g, '_');
+    // Constructing the changelog file name
+    const changelogFilePath = `./docs/changelog/v${fileName}.json`;
+    // Prepare the changelog entry
```
Comment (rarajes2) [inline packages/tools/package/src/commands/changelog/changelog.utils.ts:40]: Let's leave a blank line before (wherever we add a comment)

**Comment by sreenara** [comment]
> I don't think we need to maintain consistency with thenable syntax. We should use async-await everywhere now, it's more readable.

@rarajes2 let's stick to the conventions here since we have 4 other commands following the same design. It's easy for a person to understand the structure once they've got a hang of it.

File: .github/workflows/deploy.yml (hunk 20 above)
Comment (Shreyas281299) [reply .github/workflows/deploy.yml:262]: I think I dont think we are following any convention here. We can create a task to fix all these.

File: packages/tools/package/src/commands/changelog/changelog.utils.ts (hunk 21 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.utils.ts]: Removed

File: packages/tools/package/src/commands/changelog/changelog.utils.ts (hunk 22 above)
Comment (Shreyas281299) [reply packages/tools/package/src/commands/changelog/changelog.utils.ts:40]: Added.

**Review by rarajes2 — APPROVED** [review]
LGTM

**Review by mkesavan13 — APPROVED** [review]
Changes look good to me. No red flags. Let's merge it

**Review by sreenara — APPROVED** [review]
